File: swift/Tests/SosBroadcastServiceTests.swift
// SPDX-License-Identifier: MIT
import XCTest
@testable import AetherNetProtocol

private let LOCAL = "local"

private struct SosWireMirror: Codable {
    let broadcast_id: UUID
    let broadcast_type: String
    let message: String?
    let latitude: Double
    let longitude: Double
    let geohash: String?
}

private func newSosPacket(source: String, ttl: Int32) -> MeshPacket {
    let wire = SosWireMirror(
        broadcast_id: UUID(),
        broadcast_type: "sos",
        message: "help",
        latitude: -33.9,
        longitude: 18.4,
        geohash: nil
    )
    let payload = (try? JSONEncoder().encode(wire)) ?? Data()
    return MeshPacket(
        type: .sosBroadcast,
        sourceUhid: source,
        destinationUhid: "",
        ttl: ttl,
        priority: ProtocolConstants.sosPriority,
        payload: payload
    )
}

final class SosBroadcastServiceTests: XCTestCase {

    // MARK: - Broadcast

    func test_broadcast_floodsAndStoresAlert() async {
        let sender = FakeMeshSender(localUhid: LOCAL)
        let svc = SosBroadcastService(sender: sender)
        let ok = await svc.broadcast(broadcastType: "sos", message: "help", latitude: -33.9, longitude: 18.4)
        XCTAssertTrue(ok)
        let bcasts = sender.broadcasts()
        XCTAssertEqual(bcasts.count, 1)
        XCTAssertEqual(bcasts[0].type, .sosBroadcast)
        XCTAssertEqual(bcasts[0].ttl, ProtocolConstants.sosTtl)
        XCTAssertEqual(bcasts[0].priority, ProtocolConstants.sosPriority)
        let alerts = await svc.getActiveAlerts()
        XCTAssertEqual(alerts.count, 1)
    }

    func test_broadcast_rateLimitedAfterMax() async {
        let sender = FakeMeshSender(localUhid: LOCAL)
        let svc = SosBroadcastService(sender: sender)
        for _ in 0..<ProtocolConstants.maxSosBroadcastsPerHour {
            let ok = await svc.broadcast(broadcastType: "sos", message: "h", latitude: 0, longitude: 0)
            XCTAssertTrue(ok)
        }
        let blocked = await svc.broadcast(broadcastType: "sos", message: "h", latitude: 0, longitude: 0)
        XCTAssertFalse(blocked)
    }

    func test_broadcast_rejectsEmptyType() async {
        let sender = FakeMeshSender(localUhid: LOCAL)
        let svc = SosBroadcastService(sender: sender)
        let ok = await svc.broadcast(broadcastType: "", message: "help", latitude: 0, longitude: 0)
        XCTAssertFalse(ok)
    }

    // MARK: - Handle

    func test_handle_dropsDuplicatePacketId() async {
        let sender = FakeMeshSender(localUhid: LOCAL)
        let svc = SosBroadcastService(sender: sender)
        var pkt = newSosPacket(source: "alice", ttl: ProtocolConstants.sosTtl)
        let pktId = pkt.id

        await svc.handle(pkt)
        sender.clear()
        let alertsAfter = await svc.getActiveAlerts().count

        var pkt2 = newSosPacket(source: "alice", ttl: ProtocolConstants.sosTtl)
        pkt2.id = pktId
        await svc.handle(pkt2)

        XCTAssertEqual(sender.broadcasts().count, 0)
        let alertsNow = await svc.getActiveAlerts().count
        XCTAssertEqual(alertsNow, alertsAfter)

        // pkt is intentionally read but not mutated after first handle — discard
        // to silence the "never mutated" warning without a self-assignment.
        _ = pkt.ttl
    }

    func test_handle_ignoresSelfOriginated() async {
        let sender = FakeMeshSender(localUhid: LOCAL)
        let svc = SosBroadcastService(sender: sender)
        await svc.handle(newSosPacket(source: LOCAL, ttl: ProtocolConstants.sosTtl))
        XCTAssertEqual(sender.broadcasts().count, 0)
    }

    func test_handle_rebroadcastsWhenTtlAllows() async {
        let sender = FakeMeshSender(localUhid: LOCAL)
        let svc = SosBroadcastService(sender: sender)
        await svc.handle(newSosPacket(source: "alice", ttl: 5))
        let bcasts = sender.broadcasts()
        XCTAssertEqual(bcasts.count, 1)
        XCTAssertEqual(bcasts[0].ttl, 4)
    }

    func test_handle_doesNotRebroadcastWhenTtlExhausted() async {
        let sender = FakeMeshSender(localUhid: LOCAL)
        let svc = SosBroadcastService(sender: sender)
        await svc.handle(newSosPacket(source: "alice", ttl: 1))
        XCTAssertEqual(sender.broadcasts().count, 0)
    }

    // MARK: - Resolve

    func test_resolve_removesAlert() async {
        let sender = FakeMeshSender(localUhid: LOCAL)
        let svc = SosBroadcastService(sender: sender)
        _ = await svc.broadcast(broadcastType: "sos", message: "h", latitude: 0, longitude: 0)
        let alerts = await svc.getActiveAlerts()
        XCTAssertEqual(alerts.count, 1)
        await svc.resolve(alerts[0].id)
        let after = await svc.getActiveAlerts()
        XCTAssertTrue(after.isEmpty)
    }
}

File: tests/AetherNet.Core.Tests/SosBroadcastServiceTests.cs
// SPDX-License-Identifier: MIT

using System.Text.Json;
using AetherNet.Constants;
using AetherNet.Core.Tests.Fakes;
using AetherNet.Models;
using AetherNet.Protocol;
using AetherNet.Sos;
using Xunit;

namespace AetherNet.Core.Tests;

public class SosBroadcastServiceTests
{
    private const string Local = "local-uhid";

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
    };

    private static (SosBroadcastService svc, FakeMeshSender sender) NewService(string localUhid = Local)
    {
        var sender = new FakeMeshSender(localUhid);
        var svc = new SosBroadcastService(sender);
        return (svc, sender);
    }

    private static MeshPacket BuildSosPacketFromOther(string sourceUhid, Guid? broadcastId = null, int ttl = ProtocolConstants.SosTtl)
    {
        var body = JsonSerializer.SerializeToUtf8Bytes(new
        {
            broadcast_id = broadcastId ?? Guid.NewGuid(),
            broadcast_type = "sos",
            message = (string?)"help",
            latitude = -33.9,
            longitude = 18.4,
            geohash = (string?)null,
        }, JsonOptions);

        return new MeshPacket
        {
            Type = PacketType.SosBroadcast,
            SourceUhid = sourceUhid,
            DestinationUhid = string.Empty,
            Ttl = ttl,
            Priority = ProtocolConstants.SosPriority,
            Payload = body,
        };
    }

    // ─── BroadcastSosAsync ────────────────────────────────────

    [Fact]
    public async Task BroadcastSosAsync_FloodsAndStoresAlert()
    {
        var (svc, sender) = NewService();

        var ok = await svc.BroadcastSosAsync("sos", "help", -33.9, 18.4);

        Assert.True(ok);
        Assert.Single(sender.Broadcasts);
        var pkt = sender.Broadcasts.First();
        Assert.Equal(PacketType.SosBroadcast, pkt.Type);
        Assert.Equal(ProtocolConstants.SosTtl, pkt.Ttl);
        Assert.Equal(ProtocolConstants.SosPriority, pkt.Priority);
        Assert.Single(svc.GetActiveAlerts());
    }

    [Fact]
    public async Task BroadcastSosAsync_RateLimitedAfterMax()
    {
        var (svc, _) = NewService();

        for (var i = 0; i < ProtocolConstants.MaxSosBroadcastsPerHour; i++)
        {
            Assert.True(await svc.BroadcastSosAsync("sos", "help", 0, 0));
        }

        // Next one in the same rolling hour must be rejected.
        Assert.False(await svc.BroadcastSosAsync("sos", "help", 0, 0));
    }

    [Fact]
    public async Task BroadcastSosAsync_RejectsEmptyBroadcastType()
    {
        var (svc, _) = NewService();

        await Assert.ThrowsAnyAsync<ArgumentException>(() =>
            svc.BroadcastSosAsync(string.Empty, "help", 0, 0));
    }

    // ─── HandleAsync ──────────────────────────────────────────

    [Fact]
    public async Task HandleAsync_DropsDuplicatePacketId()
    {
        var (svc, sender) = NewService();
        var pkt = BuildSosPacketFromOther("alice");

        await svc.HandleAsync(pkt);
        sender.Clear();
        var alertsAfterFirst = svc.GetActiveAlerts().Count;

        await svc.HandleAsync(pkt);

        // Same id => no re-broadcast and no extra alert.
        Assert.Empty(sender.Broadcasts);
        Assert.Equal(alertsAfterFirst, svc.GetActiveAlerts().Count);
    }

    [Fact]
    public async Task HandleAsync_IgnoresSelfOriginated()
    {
        var (svc, sender) = NewService();
        var pkt = BuildSosPacketFromOther(Local);

        await svc.HandleAsync(pkt);

        Assert.Empty(sender.Broadcasts);
    }

    [Fact]
    public async Task HandleAsync_RaisesSosReceived()
    {
        var (svc, _) = NewService();
        SosAlert? observed = null;
        svc.SosReceived += (_, alert) => observed = alert;

        var pkt = BuildSosPacketFromOther("alice");
        await svc.HandleAsync(pkt);

        Assert.NotNull(observed);
        Assert.Equal("alice", observed!.SenderUhid);
        Assert.Equal("sos", observed.BroadcastType);
    }

    [Fact]
    public async Task HandleAsync_RebroadcastsWhenTtlAllows()
    {
        var (svc, sender) = NewService();
        var pkt = BuildSosPacketFromOther("alice", ttl: 5);

        await svc.HandleAsync(pkt);

        Assert.Single(sender.Broadcasts);
        var fwd = sender.Broadcasts.First();
        Assert.Equal(4, fwd.Ttl); // decremented before re-broadcast
    }

    [Fact]
    public async Task HandleAsync_DoesNotRebroadcastWhenTtlExhausted()
    {
        var (svc, sender) = NewService();
        var pkt = BuildSosPacketFromOther("alice", ttl: 1);

        await svc.HandleAsync(pkt);

        Assert.Empty(sender.Broadcasts);
    }

    [Fact]
    public async Task HandleAsync_RejectsWrongPacketType()
    {
        var (svc, _) = NewService();
        var pkt = new MeshPacket
        {
            Type = PacketType.Data,
            SourceUhid = "alice",
        };

        await Assert.ThrowsAsync<ArgumentException>(() => svc.HandleAsync(pkt));
    }

    // ─── Resolve ──────────────────────────────────────────────

    [Fact]
    public async Task ResolveAsync_RemovesAlertAndRaisesEvent()
    {
        var (svc, _) = NewService();
        Guid? resolvedId = null;
        svc.SosResolved += (_, id) => resolvedId = id;

        await svc.BroadcastSosAsync("sos", "help", 0, 0);
        var alert = svc.GetActiveAlerts().Single();

        await svc.ResolveAsync(alert.Id);

        Assert.Empty(svc.GetActiveAlerts());
        Assert.Equal(alert.Id, resolvedId);
    }

    [Fact]
    public async Task ResolveAsync_UnknownIdIsNoOp()
    {
        var (svc, _) = NewService();
        var raised = false;
        svc.SosResolved += (_, _) => raised = true;

        await svc.ResolveAsync(Guid.NewGuid());

        Assert.False(raised);
    }
}
